autogpt/core/messaging/base.py

```diff
@@ -1,73 +1,193 @@
-from abc import ABC, abstractmethod
-from enum import Enum, StrEnum
-import dataclasses
-from typing import Any, List, Callable, Union
-
-
-class MessageCategory(StrEnum):
-    COMMAND = "command"
-    AGENT_INSTANTIATION = "agent_instantiation"
-    USER_INPUT = "user_input"
-    USER_PROMPT = "user_prompt"
-    AGENT_MESSAGE = "agent_message"
-    SELF_FEEDBACK = "self_feedback"
-    PLAN = "plan"
+"""Abstract interface for a message broker system."""
+import abc
+from dataclasses import dataclass
+from datetime import datetime
+from enum import StrEnum
+from typing import Callable, Dict
 
 
 class Role(StrEnum):
+    """Enumeration of all entities capable of sending messages"""
+
     USER = "user"
-    SYSTEM = "system"
-    ASSISTANT = "assistant"
+    CLIENT_APPLICATION = "client_application"
+    AGENT_FACTORY = "agent_factory"
+    AGENT = "agent"
 
 
+@dataclass
 class Sender:
-    id: int  # Who is sending the message, should use interface Sender
+    """Struct for metadata about a sender."""
+
+    # We'll need some kind of uuids at some point.  Names are fine for now.
+    name: str
     role: Role
 
 
-class Receiver:
-    id: int  # Who is receiving the message, should use interface receiver
-    role: Role
+@dataclass
+class MessageMetadata:
+    """Struct for metadata about a message."""
+
+    sender: Sender
+    timestamp: datetime
+    additional_metadata: Dict = None
+    # TODO: what else goes here?
+
```

Comment:
priority: Priority

Reply:
How would this work with a normal pubsub FIFO queue?